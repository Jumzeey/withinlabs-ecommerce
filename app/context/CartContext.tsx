'use client';

import { createContext, useContext, useEffect, useState } from 'react';

interface CartItem {
  productId: string;
  quantity: number;
}

interface CartContextType {
  items: CartItem[];
  addToCart: (productId: string) => void;
  removeFromCart: (productId: string) => void;
  updateQuantity: (productId: string, quantity: number) => void;
  clearCart: () => void;
  totalItems: number;
  isInCart: (productId: string) => boolean;
}

const CART_STORAGE_KEY = 'cart';
const MAX_QUANTITY = 99;

const CartContext = createContext<CartContextType | undefined>(undefined);

export function CartProvider({ children }: { children: React.ReactNode }) {
  const [items, setItems] = useState<CartItem[]>([]);
  const [isInitialized, setIsInitialized] = useState(false);

  // Load cart from localStorage
  useEffect(() => {
    const loadCart = () => {
      try {
        const savedCart = localStorage.getItem(CART_STORAGE_KEY);
        if (savedCart) {
          const parsed = JSON.parse(savedCart);
          if (Array.isArray(parsed)) {
            setItems(parsed);
          }
        }
      } catch (err) {
        console.error('Failed to parse cart data', err);
      } finally {
        setIsInitialized(true);
      }
    };

    loadCart();
  }, []);

  // Save cart to localStorage
  useEffect(() => {
    if (isInitialized) {
      localStorage.setItem(CART_STORAGE_KEY, JSON.stringify(items));
    }
  }, [items, isInitialized]);

  const addToCart = (productId: string) => {
    setItems((prev) => {
      const existingItem = prev.find((item) => item.productId === productId);
      if (existingItem) {
        return prev.map((item) =>
          item.productId === productId
            ? { ...item, quantity: Math.min(item.quantity + 1, MAX_QUANTITY) }
            : item
        );
      }
      return [...prev, { productId, quantity: 1 }];
    });
  };

  const removeFromCart = (productId: string) => {
    setItems((prev) => prev.filter((item) => item.productId !== productId));
  };

  const updateQuantity = (productId: string, quantity: number) => {
    if (quantity < 1) {
      removeFromCart(productId);
      return;
    }

    setItems((prev) =>
      prev.map((item) =>
        item.productId === productId
          ? { ...item, quantity: Math.min(quantity, MAX_QUANTITY) }
          : item
      )
    );
  };

  const clearCart = () => {
    setItems([]);
  };

  const isInCart = (productId: string) => {
    return items.some(item => item.productId === productId);
  };

  const totalItems = items.reduce((sum, item) => sum + item.quantity, 0);

  // Only render children after cart is initialized
  if (!isInitialized) {
    return null;
  }

  return (
    <CartContext.Provider
      value={{
        items,
        addToCart,
        removeFromCart,
        updateQuantity,
        clearCart,
        totalItems,
        isInCart,
      }}
    >
      {children}
    </CartContext.Provider>
  );
}

export function useCart() {
  const context = useContext(CartContext);
  if (context === undefined) {
    throw new Error('useCart must be used within a CartProvider');
  }
  return context;
}