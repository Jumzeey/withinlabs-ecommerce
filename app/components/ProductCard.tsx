'use client';

import { Product } from '@/app/types';
import { motion } from 'framer-motion';
import { ShoppingCart } from 'lucide-react';
import Image from 'next/image';
import Link from 'next/link';
import { useCart } from '../context/CartContext';
import { Button } from '@/components/ui/button';
import { Card } from '@/components/ui/card';

export function ProductCard({ product }: { product: Product }) {
  const { addToCart } = useCart();

  return (
    <motion.div
      initial={{ opacity: 0, y: 20 }}
      animate={{ opacity: 1, y: 0 }}
      transition={{ duration: 0.3 }}
    >
      <Card className="group overflow-hidden">
        <Link href={`/products/${product.id}`}>
          <div className="relative aspect-square overflow-hidden">
            <Image
              src={product.images[0]}
              alt={product.title}
              fill
              className="object-cover transition-transform duration-300 group-hover:scale-110"
              sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
            />
          </div>
          <div className="p-4">
            <h3 className="text-lg font-semibold line-clamp-1">{product.title}</h3>
            <p className="mt-2 text-2xl font-bold text-primary">
              ${product.price.toFixed(2)}
            </p>
          </div>
        </Link>
        <div className="p-4 pt-0">
          <Button
            onClick={() => addToCart(product.id)}
            className="w-full gap-2 transition-colors"
          >
            <ShoppingCart className="h-4 w-4" />
            Add to Cart
          </Button>
        </div>
      </Card>
    </motion.div>
  );
}