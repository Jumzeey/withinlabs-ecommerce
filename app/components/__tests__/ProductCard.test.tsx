import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { ProductCard } from '../ProductCard';
import { useCart } from '../../context/CartContext';

// Mock the required modules
jest.mock('framer-motion', () => ({
  motion: {
    div: ({ children, ...props }: any) => <div {...props}>{children}</div>,
  },
}));

jest.mock('lucide-react', () => ({
  ShoppingCart: () => <svg data-testid="shopping-cart-icon" />,
}));

jest.mock('../../context/CartContext', () => ({
  useCart: jest.fn(),
}));

jest.mock('next/image', () => ({
  __esModule: true,
  default: ({ src, alt, fill, ...props }: any) => (
    // eslint-disable-next-line @next/next/no-img-element
    <img
      src={src}
      alt={alt}
      {...props}
      data-fill={fill?.toString()}
    />
  ),
}));

jest.mock('next/link', () => ({
  __esModule: true,
  default: ({ children, href, ...props }: any) => (
    <a href={href} {...props}>
      {children}
    </a>
  ),
}));

// Mock product data
const mockProduct = {
  id: 'prod-1',
  title: 'Test Product',
  price: 29.99,
  images: ['/test-image.jpg'],
  description: 'Test description',
  category: 'electronics',
  specifications: { weight: '1kg' },
  reviews: [],
};

describe('ProductCard', () => {
  const mockAddToCart = jest.fn();

  beforeEach(() => {
    (useCart as jest.Mock).mockReturnValue({
      addToCart: mockAddToCart,
    });
    jest.clearAllMocks();
  });

  it('renders product information correctly', () => {
    render(<ProductCard product={mockProduct} />);

    expect(screen.getByText(mockProduct.title)).toBeInTheDocument();
    expect(
      screen.getByText(`$${mockProduct.price.toFixed(2)}`)
    ).toBeInTheDocument();
    expect(screen.getByRole('img')).toHaveAttribute(
      'src',
      mockProduct.images[0]
    );
    expect(screen.getByRole('img')).toHaveAttribute(
      'alt',
      mockProduct.title
    );
  });

  it('contains a link to the product detail page', () => {
    render(<ProductCard product={mockProduct} />);

    const link = screen.getByRole('link');
    expect(link).toHaveAttribute('href', `/products/${mockProduct.id}`);
    // Verify the link contains the product image and title
    expect(link).toContainElement(screen.getByRole('img'));
    expect(link).toContainElement(screen.getByText(mockProduct.title));
  });

  it('renders the add to cart button with icon', () => {
    render(<ProductCard product={mockProduct} />);

    const button = screen.getByRole('button', { name: /add to cart/i });
    expect(button).toBeInTheDocument();
    expect(button).toContainElement(screen.getByTestId('shopping-cart-icon'));
  });

  it('calls addToCart when the button is clicked', () => {
    render(<ProductCard product={mockProduct} />);

    const button = screen.getByRole('button', { name: /add to cart/i });
    fireEvent.click(button);

    expect(mockAddToCart).toHaveBeenCalledTimes(1);
    expect(mockAddToCart).toHaveBeenCalledWith(mockProduct.id);
  });

  it('applies hover effect classes', () => {
    render(<ProductCard product={mockProduct} />);
    const card = screen.getByRole('link').closest('div');
    expect(card).toHaveClass('group');
    expect(screen.getByRole('img')).toHaveClass('group-hover:scale-110');
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<ProductCard product={mockProduct} />);
    expect(asFragment()).toMatchSnapshot();
  });
});