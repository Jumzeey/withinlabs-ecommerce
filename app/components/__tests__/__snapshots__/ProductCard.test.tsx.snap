// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard matches snapshot 1`] = `
<DocumentFragment>
  <div
    animate="[object Object]"
    initial="[object Object]"
    transition="[object Object]"
  >
    <div
      class="rounded-lg border bg-card text-card-foreground shadow-sm group overflow-hidden"
    >
      <a
        href="/products/prod-1"
      >
        <div
          class="relative aspect-square overflow-hidden"
        >
          <img
            alt="Test Product"
            class="object-cover transition-transform duration-300 group-hover:scale-110"
            data-fill="true"
            sizes="(max-width: 768px) 100vw, (max-width: 1200px) 50vw, 33vw"
            src="/test-image.jpg"
          />
        </div>
        <div
          class="p-4"
        >
          <h3
            class="text-lg font-semibold line-clamp-1"
          >
            Test Product
          </h3>
          <p
            class="mt-2 text-2xl font-bold text-primary"
          >
            $29.99
          </p>
        </div>
      </a>
      <div
        class="p-4 pt-0"
      >
        <button
          class="inline-flex items-center justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 bg-primary text-primary-foreground hover:bg-primary/90 h-10 px-4 py-2 w-full gap-2 transition-colors"
        >
          <svg
            data-testid="shopping-cart-icon"
          />
          Add to Cart
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
